# Report Hotel Reviews
## Preparazione e pulizia dei dati

In [ ]:
import pandas as pd # importo le librerie
import numpy as np

In [ ]:
dfhrev=pd.read_csv('Hotel_Reviews.csv') # importo il dataset

### Origine e descrizione del dataset

#### Acknowledgements

I dati sono stati estratti da Booking.com. Tutti i dati contenuti nel file sono già pubblicamente disponibili. I dati sono di proprietà di Booking.com.

#### Contenuto

Il dataset contiene circa 515.000 recensioni e valutazioni dei clienti relative a 1493 hotel di lusso in tutta Europa. Sono presenti 17 colonne, descritte di seguito:
- Hotel_Address: Indirizzo dell'hotel;

- Review_Date: Data in cui il cliente ha pubblicato la recensione corrispondente;

- Average_Score: Punteggio medio dell'hotel;

- Hotel_Name: Nome dell'hotel;

- Reviewer_Nationality: Nazionalità del cliente;

- Negative_Review: Recensione negativa;

- Review_Total_Negative_Word_Counts: Numero totale di parole nella recensione negativa;

- Positive_Review: Recensione positiva;

- Review_Total_Positive_Word_Counts: Numero totale di parole nella recensione positiva;

- Reviewer_Score: Punteggio assegnato dal cliente all'hotel;

- Total_Number_Of_Reviews_Reviewer_Has_Given (in seguito rinominato in Reviewer_Review_Count): Numero di recensioni scritte dal cliente in passato;

- Total_Number_Of_Reviews: Numero totale di recensioni dell'hotel;

- Tags: Tag assegnati dal cliente all'hotel;

- days_since_review: Durata tra la data della recensione e la data di acquisizione dei dati;

- Additional_Number_of_Scoring: Alcuni ospiti hanno assegnato un punteggio al servizio senza una recensione. Questo numero indica quanti punteggi senza recensione sono presenti;

- lat: Latitudine dell'hotel;

- lng: Longitudine dell'hotel.

### EDA

In [3]:
dfhrev.shape # restituisce il numero di righe e di colonne

(515738, 17)

In [4]:
dfhrev.head() # osservo le prime 5 righe del dataset

,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
0,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,8/3/2017,7.7,Hotel Arena,Russia,I am so angry that i made this post available...,397,1403,Only the park outside of the hotel was beauti...,11,7,2.9,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968
1,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,8/3/2017,7.7,Hotel Arena,Ireland,No Negative,0,1403,No real complaints the hotel was great great ...,105,7,7.5,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968
2,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/31/2017,7.7,Hotel Arena,Australia,Rooms are nice but for elderly a bit difficul...,42,1403,Location was good and staff were ok It is cut...,21,9,7.1,"[' Leisure trip ', ' Family with young childre...",3 days,52.360576,4.915968
3,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/31/2017,7.7,Hotel Arena,United Kingdom,My room was dirty and I was afraid to walk ba...,210,1403,Great location in nice surroundings the bar a...,26,1,3.8,"[' Leisure trip ', ' Solo traveler ', ' Duplex...",3 days,52.360576,4.915968
4,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/24/2017,7.7,Hotel Arena,New Zealand,You When I booked with your company on line y...,140,1403,Amazing location and building Romantic setting,8,3,6.7,"[' Leisure trip ', ' Couple ', ' Suite ', ' St...",10 days,52.360576,4.915968


In [5]:
dfhrev["Positive_Review"].head

<bound method NDFrame.head of 0          Only the park outside of the hotel was beauti...
1          No real complaints the hotel was great great ...
2          Location was good and staff were ok It is cut...
3          Great location in nice surroundings the bar a...
4           Amazing location and building Romantic setting 
                                ...                        
515733                                             location
515734        Breakfast was ok and we got earlier check in 
515735                                          No Positive
515736     The rooms are enormous and really comfortable...
515737                                 staff was very kind 
Name: Positive_Review, Length: 515738, dtype: object>

In [6]:
dfhrev.describe(include="all") # osservo i principali indicatori numerici incluso le colonne con valori nulli

,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
count,515738,515738.000000,515738,515738.000000,515738,515738,515738,515738.000000,515738.000000,515738,515738.000000,515738.000000,515738.000000,515738,515738,512470.000000,512470.000000
unique,1493,NaN,731,NaN,1492,227,330011,NaN,NaN,412601,NaN,NaN,NaN,55242,731,NaN,NaN
top,163 Marsh Wall Docklands Tower Hamlets London ...,NaN,8/2/2017,NaN,Britannia International Hotel Canary Wharf,United Kingdom,No Negative,NaN,NaN,No Positive,NaN,NaN,NaN,"[' Leisure trip ', ' Couple ', ' Double Room '...",1 days,NaN,NaN
freq,4789,NaN,2585,NaN,4789,245246,127890,NaN,NaN,35946,NaN,NaN,NaN,5101,2585,NaN,NaN
mean,NaN,498.081836,NaN,8.397487,NaN,NaN,NaN,18.539450,2743.743944,NaN,17.776458,7.166001,8.395077,NaN,NaN,49.442439,2.823803
std,NaN,500.538467,NaN,0.548048,NaN,NaN,NaN,29.690831,2317.464868,NaN,21.804185,11.040228,1.637856,NaN,NaN,3.466325,4.579425
min,NaN,1.000000,NaN,5.200000,NaN,NaN,NaN,0.000000,43.000000,NaN,0.000000,1.000000,2.500000,NaN,NaN,41.328376,-0.369758
25%,NaN,169.000000,NaN,8.100000,NaN,NaN,NaN,2.000000,1161.000000,NaN,5.000000,1.000000,7.500000,NaN,NaN,48.214662,-0.143372
50%,NaN,341.000000,NaN,8.400000,NaN,NaN,NaN,9.000000,2134.000000,NaN,11.000000,3.000000,8.800000,NaN,NaN,51.499981,0.010607
75%,NaN,660.000000,NaN,8.800000,NaN,NaN,NaN,23.000000,3613.000000,NaN,22.000000,8.000000,9.600000,NaN,NaN,51.516288,4.834443


In [7]:
dfhrev.describe() # osservo i principali indicatori numerici escludendo le colonne con valori nulli

,Additional_Number_of_Scoring,Average_Score,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,lat,lng
count,515738.000000,515738.000000,515738.000000,515738.000000,515738.000000,515738.000000,515738.000000,512470.000000,512470.000000
mean,498.081836,8.397487,18.539450,2743.743944,17.776458,7.166001,8.395077,49.442439,2.823803
std,500.538467,0.548048,29.690831,2317.464868,21.804185,11.040228,1.637856,3.466325,4.579425
min,1.000000,5.200000,0.000000,43.000000,0.000000,1.000000,2.500000,41.328376,-0.369758
25%,169.000000,8.100000,2.000000,1161.000000,5.000000,1.000000,7.500000,48.214662,-0.143372
50%,341.000000,8.400000,9.000000,2134.000000,11.000000,3.000000,8.800000,51.499981,0.010607
75%,660.000000,8.800000,23.000000,3613.000000,22.000000,8.000000,9.600000,51.516288,4.834443
max,2682.000000,9.800000,408.000000,16670.000000,395.000000,355.000000,10.000000,52.400181,16.429233


In [8]:
dfhrev.dtypes # visualizzo il tipo di dato

Hotel_Address                                  object
Additional_Number_of_Scoring                    int64
Review_Date                                    object
Average_Score                                 float64
Hotel_Name                                     object
Reviewer_Nationality                           object
Negative_Review                                object
Review_Total_Negative_Word_Counts               int64
Total_Number_of_Reviews                         int64
Positive_Review                                object
Review_Total_Positive_Word_Counts               int64
Total_Number_of_Reviews_Reviewer_Has_Given      int64
Reviewer_Score                                float64
Tags                                           object
days_since_review                              object
lat                                           float64
lng                                           float64
dtype: object

In [9]:
dfhrev.info() # descrizione del dataframe in cui, oltre al tipo di dato, si può osservare il numero di celle con valori non nulli

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 515738 entries, 0 to 515737
Data columns (total 17 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   Hotel_Address                               515738 non-null  object 
 1   Additional_Number_of_Scoring                515738 non-null  int64  
 2   Review_Date                                 515738 non-null  object 
 3   Average_Score                               515738 non-null  float64
 4   Hotel_Name                                  515738 non-null  object 
 5   Reviewer_Nationality                        515738 non-null  object 
 6   Negative_Review                             515738 non-null  object 
 7   Review_Total_Negative_Word_Counts           515738 non-null  int64  
 8   Total_Number_of_Reviews                     515738 non-null  int64  
 9   Positive_Review                             515738 non-null  object 
 

In [10]:
dfhrev.isnull().sum() # restituisce il numero di celle con valori nulli per ciascuna colonna

Hotel_Address                                    0
Additional_Number_of_Scoring                     0
Review_Date                                      0
Average_Score                                    0
Hotel_Name                                       0
Reviewer_Nationality                             0
Negative_Review                                  0
Review_Total_Negative_Word_Counts                0
Total_Number_of_Reviews                          0
Positive_Review                                  0
Review_Total_Positive_Word_Counts                0
Total_Number_of_Reviews_Reviewer_Has_Given       0
Reviewer_Score                                   0
Tags                                             0
days_since_review                                0
lat                                           3268
lng                                           3268
dtype: int64

In [11]:
dfhrev[["Reviewer_Score","Average_Score"]].describe() # osservo la distribuzione di Reviewer_Score e Average_Score


,Reviewer_Score,Average_Score
count,515738.000000,515738.000000
mean,8.395077,8.397487
std,1.637856,0.548048
min,2.500000,5.200000
25%,7.500000,8.100000
50%,8.800000,8.400000
75%,9.600000,8.800000
max,10.000000,9.800000


Dall'esplorazione del DataFrame si evince che le uniche colonne ad avere valori nulli sono lat e lng, entrambe con 3268 valori mancanti. Inoltre, osservando le distribuzioni di Reviewer Score ed Average Score si nota che i valori hanno un range compreso tra 0 e 10, come atteso, e che la loro media si attesta intorno ad 8.3, evidenziando una maggioranza di punteggi elevati.

In [12]:
dfhrev2=dfhrev.copy() # creo una copia del DataFrame da utilizzare per le successive analisi, in modo da lasciare intatto il DataFrame originale 

### Data Cleaning

#### Gestione missing values
Dopo aver verificato la presenza di valori mancanti ed aver appurato che sono presenti soltanto nelle colonne lat e lng, procedo alla loro eliminazione in quanto rispetto al totale di oltre 500k righe possono essere trascurati.

In [13]:
# Gestione missing values
dfhrev2.dropna(subset=['lat', 'lng'], inplace=True) # elimino righe con valori nulli nelle colonne lat e lng
dfhrev2.isnull().sum() # verifico di aver eliminato i valori nulli

Hotel_Address                                 0
Additional_Number_of_Scoring                  0
Review_Date                                   0
Average_Score                                 0
Hotel_Name                                    0
Reviewer_Nationality                          0
Negative_Review                               0
Review_Total_Negative_Word_Counts             0
Total_Number_of_Reviews                       0
Positive_Review                               0
Review_Total_Positive_Word_Counts             0
Total_Number_of_Reviews_Reviewer_Has_Given    0
Reviewer_Score                                0
Tags                                          0
days_since_review                             0
lat                                           0
lng                                           0
dtype: int64

#### Conversione in datetime ed estrazione anno/mese
Avendo osservato che la colonna Review_Date è di tipo object, la converto in datetime per una maggiore accuratezza. Inoltre, per facilitare analisi successive creo delle nuove colonne dove inserire esclusivamente anno, mese ed anno e mese insieme, quest'ultima utile per poter raggruppare le recensioni in base ad un periodo mensile.

In [14]:
# Conversione in datetime
dfhrev2['Review_Date'] = pd.to_datetime(dfhrev2['Review_Date']) # converto la colonna Review_Date da tipo object a datetime
print(dfhrev2['Review_Date'].dtype) # controllo che il dato sia stato convertito correttamente

datetime64[ns]


In [15]:
# Estrazione anno e mese creando nuove colonne
dfhrev2['Review_Year'] = dfhrev2['Review_Date'].dt.year # estraggo l'anno 
dfhrev2['Review_Month'] = dfhrev2['Review_Date'].dt.month # estraggo il mese
dfhrev2['Review_YearMonth'] = dfhrev2['Review_Date'].dt.to_period('M') # estraggo anno e mese insieme creando un periodo temporale mensile, dove 'M' = Monthly (mese)
print("Estrazione anno e mese:") # controllo le nuove colonne
print(dfhrev2[['Review_Date', 'Review_Year', 'Review_Month', 'Review_YearMonth']].head())

Estrazione anno e mese:
  Review_Date  Review_Year  Review_Month Review_YearMonth
0  2017-08-03         2017             8          2017-08
1  2017-08-03         2017             8          2017-08
2  2017-07-31         2017             7          2017-07
3  2017-07-31         2017             7          2017-07
4  2017-07-24         2017             7          2017-07


#### Rimozione duplicati esatti
Prima di procedere alla rimozione dei duplicati esatti, conto il numero di righe presenti, corrispondente a 512470. Dopodiché, procedo alla rimozione dei duplicati attraverso drop_duplicates, che identifica e rimuove righe completamente identiche su tutte le colonne, mantenendo la prima occorrenza ed eliminando le successive. Questo passaggio è utile per evitare errori nelle analisi successive o ridondanza dei dati.

In [16]:
print("Numero di righe:", len(dfhrev2)) # conto il numero di righe

Numero di righe: 512470


In [17]:
dfhrev2 = dfhrev2.drop_duplicates() # rimuovo i duplicati
print("Numero di righe dopo rimozione:", len(dfhrev2)) # conto il numero di righe dopo la rimozione dei duplicati


Numero di righe dopo rimozione: 511944


In [18]:
print("Duplicati rimossi:", 512470 - len(dfhrev2)) # osservo il numero di righe rimosse

Duplicati rimossi: 526


#### Creazione variabili derivate
Per definire la lunghezza totale di ciascuna recensione, creo una nuova colonna, chiamata Review_Length, in cui sommare le parole positive e negative presenti nelle recensioni. Inoltre, per maggiore chiarezza e comodità, rinomino la colonna Total_Number_of_Reviews_Reviewer_Has_Given in Reviewer_Review_Count. 

In [19]:
# Creazione variabili derivate
dfhrev2['Review_Length'] = (dfhrev2['Review_Total_Negative_Word_Counts'] + 
                                 dfhrev2['Review_Total_Positive_Word_Counts']) # creo una nuova colonna per definire la lunghezza totale della recensione
dfhrev2['Reviewer_Review_Count'] = dfhrev2['Total_Number_of_Reviews_Reviewer_Has_Given'] # rinomino la colonna 



In [20]:
print("Shape finale:", dfhrev2.shape) # osservo il numero finale di righe e di colonne dopo la pulizia dei dati
print("Colonne aggiunte:", [col for col in dfhrev2.columns if col not in # controllo le colonne aggiunte al dataframe 
                           ['Hotel_Address', 'Review_Date', 'Average_Score', 'Hotel_Name', 
                            'Reviewer_Nationality', 'Negative_Review', 
                            'Review_Total_Negative_Word_Counts', 'Positive_Review', 
                            'Review_Total_Positive_Word_Counts', 'Reviewer_Score', 
                            'Total_Number_of_Reviews_Reviewer_Has_Given', 
                            'Total_Number_of_Reviews', 'Tags', 'days_since_review', 
                            'Additional_Number_of_Scoring', 'lat', 'lng']])

Shape finale: (511944, 22)
Colonne aggiunte: ['Review_Year', 'Review_Month', 'Review_YearMonth', 'Review_Length', 'Reviewer_Review_Count']


In [21]:
print("Colonne:", dfhrev2.columns.tolist())

Colonne: ['Hotel_Address', 'Additional_Number_of_Scoring', 'Review_Date', 'Average_Score', 'Hotel_Name', 'Reviewer_Nationality', 'Negative_Review', 'Review_Total_Negative_Word_Counts', 'Total_Number_of_Reviews', 'Positive_Review', 'Review_Total_Positive_Word_Counts', 'Total_Number_of_Reviews_Reviewer_Has_Given', 'Reviewer_Score', 'Tags', 'days_since_review', 'lat', 'lng', 'Review_Year', 'Review_Month', 'Review_YearMonth', 'Review_Length', 'Reviewer_Review_Count']


In [22]:
drop_cols = ['Total_Number_of_Reviews_Reviewer_Has_Given'] # dopo aver visualizzato le colonne, decido di eliminare la colonna iniziale in quanto identica a Reviewer_Review_Count 
dfhrev2.drop(columns=drop_cols, inplace=True, errors='ignore')
# Verifica shape e colonne
print("dfhrev2 pulito:", dfhrev2.shape) # controllo il dataframe finale
print("Colonne:", dfhrev2.columns.tolist())

dfhrev2 pulito: (511944, 21)
Colonne: ['Hotel_Address', 'Additional_Number_of_Scoring', 'Review_Date', 'Average_Score', 'Hotel_Name', 'Reviewer_Nationality', 'Negative_Review', 'Review_Total_Negative_Word_Counts', 'Total_Number_of_Reviews', 'Positive_Review', 'Review_Total_Positive_Word_Counts', 'Reviewer_Score', 'Tags', 'days_since_review', 'lat', 'lng', 'Review_Year', 'Review_Month', 'Review_YearMonth', 'Review_Length', 'Reviewer_Review_Count']


In [23]:
# Salvataggio del nuovo DataFrame pulito da utilizzare per SQL
dfhrev2.to_csv('hotel_reviews_clean.csv', index=False)
print("Salvataggio hotel_reviews_clean.csv")

Salvataggio hotel_reviews_clean.csv


#### Preparazione DataFrame per popolare il DataBase in SQL dividendolo in 4 CSV

##### Creazione tabella hotels

Verifica nomi hotel e indirizzi:

In [50]:
print(len(dfhrev2["Hotel_Name"].unique())) # controllo che il numero di hotel combaci con quello degli indirizzi
print(len(dfhrev2["Hotel_Address"].unique()))

1475
1476


In [25]:
# Risulta esserci un indirizzo in più (1476) rispetto al nome degli hotel (1475), procedo ad identificare l'anomalia
anomalie = dfhrev2.groupby('Hotel_Address')['Hotel_Name'].nunique().reset_index()
anomalie = anomalie[anomalie['Hotel_Name'] > 1].rename(columns={'Hotel_Name': 'n_nomi'})
print("Anomalie trovate:")
print(anomalie)

# Dettaglio casi specifici
for idx, row in anomalie.iterrows():
    indirizzo = row['Hotel_Address']
    nomi = dfhrev2[dfhrev2['Hotel_Address'] == indirizzo]['Hotel_Name'].unique()
    print(f"\n{indirizzo}:")
    print(f"  Nomi: {list(nomi)}")

# Conto totali per conferma
n_indirizzi = dfhrev2['Hotel_Address'].nunique()
n_nomi = dfhrev2['Hotel_Name'].nunique()
print(f"\nTotale: {n_indirizzi} indirizzi univoci, {n_nomi} nomi univoci")

Anomalie trovate:
                                         Hotel_Address  n_nomi
719  8 Northumberland Avenue Westminster Borough Lo...       2

8 Northumberland Avenue Westminster Borough London WC2N 5BY United Kingdom:
  Nomi: ['The Grand at Trafalgar Square', 'Club Quarters Hotel Trafalgar Square']

Totale: 1476 indirizzi univoci, 1475 nomi univoci


In [51]:
dfhrev2_clean = dfhrev2.drop_duplicates(subset=['Hotel_Name', 'Hotel_Address'])
print(f"\nDopo drop_duplicates(Nome+Indirizzo): {len(dfhrev2_clean)} righe")


Dopo drop_duplicates(Nome+Indirizzo): 1477 righe


In [27]:
# Siccome ho trovato 1477 righe anziché 1476 procedo ad un ulteriore verifica
n_righe_originale = len(dfhrev2)
n_indirizzi = dfhrev2['Hotel_Address'].nunique()
n_nomi = dfhrev2['Hotel_Name'].nunique()
n_unici_nome_indirizzo = len(dfhrev2.drop_duplicates(subset=['Hotel_Name', 'Hotel_Address']))

print(f"Righe totali: {n_righe_originale:,}")
print(f"Indirizzi univoci: {n_indirizzi}")
print(f"Nomi univoci: {n_nomi}")
print(f"Unici (Nome+Indirizzo): {n_unici_nome_indirizzo}")

print(f"\nVERIFICA")
if n_unici_nome_indirizzo == n_indirizzi:
    print("CORRETTO: drop_duplicates() = n_indirizzi")
else:
    print("Controllo dati nulli")
    
# Controlla nulli
print(f"\nNull Hotel_Address: {dfhrev2['Hotel_Address'].isnull().sum()}")
print(f"Null Hotel_Name: {dfhrev2['Hotel_Name'].isnull().sum()}")

Righe totali: 511,944
Indirizzi univoci: 1476
Nomi univoci: 1475
Unici (Nome+Indirizzo): 1477

VERIFICA
Controllo dati nulli

Null Hotel_Address: 0
Null Hotel_Name: 0


In [28]:
# Avendo trovato una discrepanza, verifico se sono presenti nomi che hanno più indirizzi
nomi_multi_indirizzo = dfhrev2.groupby('Hotel_Name')['Hotel_Address'].nunique()
nomi_multi_indirizzo = nomi_multi_indirizzo[nomi_multi_indirizzo > 1]

print("Nomi con più indirizzi:")
print(nomi_multi_indirizzo)

# Dettaglio
for nome in nomi_multi_indirizzo.index:
    indirizzi = dfhrev2[dfhrev2['Hotel_Name'] == nome]['Hotel_Address'].unique()
    print(f"\n{nome}:")
    print(f"  {list(indirizzi)}")

Nomi con più indirizzi:
Hotel_Name
Hotel Regina    3
Name: Hotel_Address, dtype: int64

Hotel Regina:
  ['Bergara 2 4 Eixample 08002 Barcelona Spain', 'Rooseveltplatz 15 09 Alsergrund 1090 Vienna Austria', 'Via Cesare Correnti 13 Milan City Center 20123 Milan Italy']


In [29]:
# Verifica finale completa
unici = dfhrev2[['Hotel_Name', 'Hotel_Address']].drop_duplicates()
print(f"Combinazioni uniche Nome+Indirizzo: {len(unici)}")

# Distribuzione per indirizzo
duplicati_indirizzo = dfhrev2.groupby('Hotel_Address')['Hotel_Name'].nunique()
print(f"Indirizzi con 1 nome: {(duplicati_indirizzo == 1).sum()}")
print(f"Indirizzi con 2+ nomi: {(duplicati_indirizzo > 1).sum()}")

# Distribuzione per nome  
duplicati_nome = dfhrev2.groupby('Hotel_Name')['Hotel_Address'].nunique()
print(f"Nomi con 1 indirizzo: {(duplicati_nome == 1).sum()}")
print(f"Nomi con 2+ indirizzi: {(duplicati_nome > 1).sum()}")

Combinazioni uniche Nome+Indirizzo: 1477
Indirizzi con 1 nome: 1475
Indirizzi con 2+ nomi: 1
Nomi con 1 indirizzo: 1474
Nomi con 2+ indirizzi: 1


In [30]:
# Procedo ad assegnare un id univoco agli hotel da utilizzare per le successive analisi in SQL utilizzando nome+indirizzo (1477 ID)
hotels_df = (dfhrev2_clean[['Hotel_Name', 'Hotel_Address', 'lat', 'lng']]
             .drop_duplicates() 
             .reset_index(drop=True)
             .assign(hotel_id=lambda x: x.index))  # 0,1,2,...,1476

# Rinomina per SQL 
hotels_df = hotels_df.rename(columns={
    'Hotel_Name': 'hotel_name',
    'Hotel_Address': 'hotel_address'
})[['hotel_id', 'hotel_name', 'hotel_address', 'lat', 'lng']]

print(f"Hotels: {len(hotels_df)} righe")
print("Colonne:", hotels_df.columns.tolist())
hotels_df.head()

Hotels: 1477 righe
Colonne: ['hotel_id', 'hotel_name', 'hotel_address', 'lat', 'lng']


,hotel_id,hotel_name,hotel_address,lat,lng
0,0,Hotel Arena,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,52.360576,4.915968
1,1,K K Hotel George,1 15 Templeton Place Earl s Court Kensington a...,51.491888,-0.194971
2,2,Apex Temple Court Hotel,1 2 Serjeant s Inn Fleet Street City of London...,51.513734,-0.108751
3,3,The Park Grand London Paddington,1 3 Queens Garden Westminster Borough London W...,51.514218,-0.180903
4,4,Monhotel Lounge SPA,1 3 Rue d Argentine 16th arr 75116 Paris France,48.874348,2.289733


#### Gestione anomalie nomi hotel ed indirizzi
L'analisi preliminare del dataset ha evidenziato due anomalie strutturali nella corrispondenza tra i nomi degli hotel e gli indirizzi.

##### Anomalia 1: Indirizzo con più nomi (1 caso)
Indirizzo: 
- "8 Northumberland Avenue Westminster Borough London WC2N 5BY"

Nomi: 
- "The Grand at Trafalgar Square"
- "Club Quarters Hotel Trafalgar Square"
Interpretazione: Probabile rebranding o gestione separata su Booking.com dello stesso immobile fisico.

Soluzione: Gli indirizzi sono stati gestiti come 2 record distinti.
Risultato: 1476 combinazioni univoche (invece di 1475 attese).

##### Anomalia 2: Nome con più indirizzi (1 caso)
Hotel Regina appare in:
- Bergara 2-4, Barcellona (Spagna)
- Rooseveltplatz 15, Vienna (Austria)
- Via Cesare Correnti 13, Milano (Italia)
Interpretazione: Catena alberghiera; "Hotel Regina" con sedi multiple in Europa.
Soluzione: Gli hotel sono stati getsiti come 3 record distinti.
Risultato: 1.477 Hotel distinti (invece di 1.475 attesi).

##### Resoconto finale
Al fine di utilizzare il DataFrame per successive analisi in SQL e la crezione di identificativi per ciascun hotel (hotel_id), si è scelto di trattare gli hotel con anomalie come entità distinte: 
1.475 hotel (1 nome = 1 indirizzo)
1 hotel multi-nome (1 indirizzo = 2 nomi)
1 catena multi-sede (1 nome = 3 indirizzi)
Risultato: 1.477 hotel_id univoci, tutte le recensioni preservate.

###### Vantaggi scelta
| Criterio | Soluzione adottata |
|----------|-------------------|
| **Integrità dati** | 100% recensioni preservate |
| **Flessibilità** | Query per indirizzo/nome/catena |
| **Realismo** | Riflette gestione Booking.com |
| **Semplicità** | Nessuna unione complessa o manuale |



Procedo con il salvataggio in CSV per l'importazione in SQL

In [31]:
hotels_df.to_csv('hotels.csv', index=False)
print("Salvataggio hotels.csv")

Salvataggio hotels.csv


Importo il file CSV direttamente tramite python (sqlalchemy) su MySQL per popolare la tabella hotels creata

In [32]:
from sqlalchemy import create_engine

csv_path = "hotels.csv"

df = pd.read_csv(csv_path)

engine = create_engine("mysql+pymysql://root:@127.0.0.1:3306/hotel_reviews")

df.to_sql(
    name="hotels",
    con=engine,
    if_exists="append", # inserisce le righe nella tabella creata in MySQL
    index=False
)

print("Import completato")


KeyboardInterrupt



In [ ]:
from sqlalchemy import create_engine

csv_path = "hotels.csv"

df = pd.read_csv(csv_path)

engine = create_engine("mysql+pymysql://root:@127.0.0.1:3306/hotel_reviews")

df.to_sql(
    name="hotels",
    con=engine,
    if_exists="append", # inserisce le righe nella tabella creata in MySQL
    index=False
)

print("Import completato")

Creo due nuove variabili derivate denominate hotel_city e nation, in modo da poter avere a disposizione i nomi delle città e delle nazioni in cui si trovano gli hotel per facilitare eventuali analisi geografiche. I nomi sono stati estrapolati a partire dagli indirizzi e salvati in files di testo .txt, utilizzati in seguito per la creazione delle nuove colonne.

In [2]:
hotels = pd.read_csv("hotels.csv")

cities = pd.read_csv("cities.txt", header=None, names=["hotel_city"])

assert len(hotels) == len(cities), "Le righe non coincidono"

hotels["hotel_city"] = cities["hotel_city"].values

In [9]:
hotels = pd.read_csv("hotels.csv")

nations = pd.read_csv("nations.txt", header=None, names=["nation"])

assert len(hotels) == len(nations), "Le righe non coincidono"

hotels["nation"] = nations["nation"].values

In [10]:
print(hotels[["hotel_address", "hotel_city", "nation"]].head(20).to_string(index=False))

                                                                         hotel_address hotel_city      nation
                              s Gravesandestraat 55 Oost 1092 AA Amsterdam Netherlands  Amsterdam Paesi Bassi
1 15 Templeton Place Earl s Court Kensington and Chelsea London SW5 9NB United Kingdom     Londra Regno Unito
         1 2 Serjeant s Inn Fleet Street City of London London EC4Y 1LL United Kingdom     Londra Regno Unito
                    1 3 Queens Garden Westminster Borough London W2 3BA United Kingdom     Londra Regno Unito
                                       1 3 Rue d Argentine 16th arr 75116 Paris France     Parigi     Francia
                                        1 5 Passage Ruelle 18th arr 75018 Paris France     Parigi     Francia
                              1 8 Russell Square Camden London WC1B 5BE United Kingdom     Londra Regno Unito
                              1 Addington Street Lambeth London SE1 7RY United Kingdom     Londra Regno Unito
          

In [11]:
hotels.to_csv("hotels.csv", index=False)

In [12]:
print(hotels.columns.tolist())

['hotel_id', 'hotel_name', 'hotel_address', 'lat', 'lng', 'hotel_city', 'nation']


Utilizzo sqlalchemy per popolare le nuove colonne hotel_city e nation con i nomi delle città e delle nazioni, ordinandole utilizzando la chiave primaria hotel_id.

In [ ]:
from sqlalchemy import create_engine, text

engine = create_engine("mysql+pymysql://root:@127.0.0.1:3306/hotel_reviews")

df = pd.read_csv("hotels.csv")
updates = df[["hotel_city", "hotel_id"]].to_dict(orient="records")

sql = text("""
    UPDATE hotels
    SET hotel_city = :hotel_city
    WHERE hotel_id = :hotel_id
""")

with engine.begin() as conn:
    conn.execute(sql, updates)

In [13]:
from sqlalchemy import create_engine, text

engine = create_engine("mysql+pymysql://root:@127.0.0.1:3306/hotel_reviews")

df = pd.read_csv("hotels.csv")
updates = df[["nation", "hotel_id"]].to_dict(orient="records")

sql = text("""
    UPDATE hotels
    SET nation = :nation
    WHERE hotel_id = :hotel_id
""")

with engine.begin() as conn:
    conn.execute(sql, updates)

##### Creazione tabella reviewer

In [2]:
dfhrev2=pd.read_csv('hotel_reviews_clean.csv')

In [34]:
reviewers_df = (
    dfhrev2[['Reviewer_Nationality', 'Reviewer_Review_Count']]
    .drop_duplicates()
    .reset_index(drop=True)
    .reset_index(names='reviewer_id')
    .rename(columns={
        'Reviewer_Nationality': 'reviewer_nationality',
        'Reviewer_Review_Count': 'reviewer_review_count'
    })[['reviewer_id', 'reviewer_nationality', 'reviewer_review_count']]
)

print(f"Reviewers: {len(reviewers_df)}")
print("Colonne:", reviewers_df.columns.tolist())
reviewers_df.head()

Reviewers: 6340
Colonne: ['reviewer_id', 'reviewer_nationality', 'reviewer_review_count']


,reviewer_id,reviewer_nationality,reviewer_review_count
0,0,Russia,7
1,1,Ireland,7
2,2,Australia,9
3,3,United Kingdom,1
4,4,New Zealand,3


In [ ]:
reviewers_df.to_csv('reviewers.csv', index=False)
print("Salvataggio reviewers.csv")

Salvataggio reviewers.csv


Importo il file reviewers su SQL

In [ ]:
# Controllo la tabella
print(reviewers_df.head())
print(reviewers_df.columns.tolist())
print(reviewers_df.shape)

   reviewer_id reviewer_nationality  reviewer_review_count
0            0              Russia                       7
1            1             Ireland                       7
2            2           Australia                       9
3            3      United Kingdom                       1
4            4         New Zealand                       3
['reviewer_id', 'reviewer_nationality', 'reviewer_review_count']
(6340, 3)


In [ ]:
from sqlalchemy import create_engine

csv_path = "reviewers.csv"

df = pd.read_csv(csv_path)

engine = create_engine("mysql+pymysql://root:@127.0.0.1:3306/hotel_reviews")

df.to_sql(
    name="reviewers",
    con=engine,
    if_exists="append",
    index=False
)

print("Import completato")

Import completato


##### Creazione tabella hotel_stats

In [ ]:
hotel_stats_df = (
    dfhrev2_clean[[
        'Hotel_Name',
        'Hotel_Address',
        'Average_Score',
        'Additional_Number_of_Scoring',
        'Total_Number_of_Reviews'
    ]]
    .drop_duplicates()
    .merge(
        hotels_df[['hotel_id', 'hotel_name', 'hotel_address']],
        left_on=['Hotel_Name', 'Hotel_Address'],
        right_on=['hotel_name', 'hotel_address'],
        how='inner'
    )
    .rename(columns={
        'Average_Score': 'average_score',
        'Additional_Number_of_Scoring': 'additional_number_of_scoring',
        'Total_Number_of_Reviews': 'total_number_of_reviews'
    })[
        [
            'hotel_id',
            'average_score',
            'additional_number_of_scoring',
            'total_number_of_reviews'
        ]
    ]
    .drop_duplicates(subset=['hotel_id'])
    .sort_values('hotel_id')
    .reset_index(drop=True)
)

hotel_stats_df.to_csv('hotel_stats.csv', index=False)

In [ ]:
hotel_stats_df.to_csv('hotel_stats.csv', index=False)
print("Salvataggio hotel_stats.csv completato")

Salvataggio hotel_stats.csv completato
None


Importo il file hotel_stats su SQL

In [ ]:
# Controllo la tabella
print(hotel_stats_df.head())
print(hotel_stats_df.columns.tolist())
print(hotel_stats_df.shape)

   hotel_id  average_score  additional_number_of_scoring  \
0         0            7.7                           194   
1         1            8.5                           244   
2         2            9.2                           639   
3         3            7.7                          1058   
4         4            8.4                            26   

   total_number_of_reviews  
0                     1403  
1                     1831  
2                     2619  
3                     4380  
4                      171  
['hotel_id', 'average_score', 'additional_number_of_scoring', 'total_number_of_reviews']
(1477, 4)


In [ ]:
from sqlalchemy import create_engine

df = pd.read_csv("hotel_stats.csv")

engine = create_engine("mysql+pymysql://root:@127.0.0.1:3306/hotel_reviews")

df.to_sql(
    name="hotel_stats",
    con=engine,
    if_exists="append",
    index=False
)

print("Import completato")

Import completato


##### Creazione tabella reviews

In [57]:
# Ripulisco l'indice: da 0 a N-1, senza "buchi" da operazioni precedenti
dfhrev2 = dfhrev2.reset_index(drop=True)

# Creo la chiave surrogata review_id come indice + 1 (parte da 1, non da 0)
dfhrev2["review_id"] = dfhrev2.index + 1

# Verifico che review_id sia effettivamente univoco
print("review_id univoco:", dfhrev2["review_id"].is_unique)
print("Numero di recensioni:", dfhrev2.shape[0])

review_id univoco: True
Numero di recensioni: 511944


Ho creato una chiave surrogata review_id assegnando un ID progressivo a ogni riga del dataset pulito. Questa chiave funge da primary key nella tabella reviews e garantisce un identificativo univoco per ciascuna recensione.

In [58]:
# merge con hotels per ricavare hotel_id 
reviews_with_hotels = dfhrev2.merge(
    hotels_df[["hotel_id", "hotel_name", "hotel_address"]],
    left_on=["Hotel_Name", "Hotel_Address"],
    right_on=["hotel_name", "hotel_address"],
    how="left"  # LEFT: mantengo tutte le recensioni, anche se qualche hotel non matcha
)

# Controllo quante recensioni non hanno trovato un hotel_id
missing_hotels = reviews_with_hotels["hotel_id"].isna().sum()
print("Recensioni senza hotel_id dopo la merge:", missing_hotels)

Recensioni senza hotel_id dopo la merge: 0


Per collegare ogni recensione al relativo hotel, ho effettuato una join tra il dataset in cui sono presenti le recensioni e hotels, utilizzando il nome e l’indirizzo dell’hotel. In questo modo ho recuperato l'hotel_id già definito nella tabella hotels e ho verificato il numero di recensioni per cui l’identificativo non è stato trovato.

In [59]:
# merge con reviewers per ricavare reviewer_id
reviews_with_ids = reviews_with_hotels.merge(
    reviewers_df[["reviewer_id", "reviewer_nationality", "reviewer_review_count"]],
    left_on=["Reviewer_Nationality", "Reviewer_Review_Count"],
    right_on=["reviewer_nationality", "reviewer_review_count"],
    how="left"  # LEFT: mantengo tutte le recensioni
)

# Controllo quante recensioni non hanno trovato un reviewer_id
missing_reviewers = reviews_with_ids["reviewer_id"].isna().sum()
print("Recensioni senza reviewer_id dopo la merge:", missing_reviewers)

Recensioni senza reviewer_id dopo la merge: 0


In modo analogo, ho unito il dataset delle recensioni con reviewers sfruttando la coppia (Reviewer_Nationality, Reviewer_Review_Count) come chiave di collegamento. Questa join mi ha permesso di associare a ogni recensione il reviewer_id già presente nella tabella dei reviewers.

In [60]:
reviews_df = reviews_with_ids[[
    "review_id",
    "hotel_id",
    "reviewer_id",
    "Review_Date",
    "Review_Year",
    "Review_Month",
    "Review_YearMonth",
    "Review_Length",
    "days_since_review",
    "Negative_Review",
    "Positive_Review",
    "Review_Total_Positive_Word_Counts",
    "Review_Total_Negative_Word_Counts",
    "Reviewer_Score"
]].rename(columns={
    "Review_Date": "review_date",
    "Review_Year": "review_year",
    "Review_Month": "review_month",
    "Review_YearMonth": "review_yearmonth",
    "Review_Length": "review_length",
    "Negative_Review": "negative_review",
    "Positive_Review": "positive_review",
    "Review_Total_Positive_Word_Counts": "review_total_positive_word_counts",
    "Review_Total_Negative_Word_Counts": "review_total_negative_word_counts",
    "Reviewer_Score": "reviewer_score"
})

print("Shape finale reviews_df:", reviews_df.shape)
print(reviews_df.head())

Shape finale reviews_df: (511944, 14)
   review_id  hotel_id  reviewer_id review_date  review_year  review_month  \
0          1         0            0  2017-08-03         2017             8   
1          2         0            1  2017-08-03         2017             8   
2          3         0            2  2017-07-31         2017             7   
3          4         0            3  2017-07-31         2017             7   
4          5         0            4  2017-07-24         2017             7   

  review_yearmonth  review_length days_since_review  \
0          2017-08            408            0 days   
1          2017-08            105            0 days   
2          2017-07             63            3 days   
3          2017-07            236            3 days   
4          2017-07            148           10 days   

                                     negative_review  \
0   I am so angry that i made this post available...   
1                                        No Negati

A partire dal dataset arricchito con le chiavi esterne, ho costruito il dataframe finale reviews_df selezionando esclusivamente gli attributi necessari per la tabella reviews e rinominando le colonne in quanto in sql i nomi sono riportati tutti in minuscolo. Il risultato è una tabella con una riga per recensione, primary key review_id e foreign key hotel_id e reviewer_id.

In [62]:
output_csv_path = "reviews.csv"
reviews_df.to_csv(output_csv_path, index=False)

print(f"File CSV creato: {output_csv_path}")

File CSV creato: reviews.csv


Infine, ho esportato il dataframe reviews_df in un file reviews.csv utilizzato di seguito per popolare la tabella reviews nel database in  MySQL.

Importo il file reviews in SQL

In [63]:
from sqlalchemy import create_engine

df = pd.read_csv("reviews.csv")

engine = create_engine("mysql+pymysql://root:@127.0.0.1:3306/hotel_reviews")

df.to_sql(
    name="reviews",
    con=engine,
    if_exists="append",
    index=False
)

print("Import completato")

Import completato
